### DISCUSSION TITLE
Feature: Add Generated Item to Barrel File - Then (optionally) Import in Module

### DISCUSSION THREAD

**Discussion by andy-bond** [cyrilletuzi/vscode-angular-schematics#419] [Ideas]
# Checks before posting an issue

- [x] I have read all the extension's documentation (README and docs/TROUBLESHOOTING)
- [x] My project meets the requirements indicated in README
- [x] I launched `ng g component hello` in the *VS Code* Terminal inside my project, and it works
- [x] I looked at the extension's logs ("Angular Schematics" Output channel, second tab left to the Terminal), and tried to fix my project's configuration accordingly
- [x] I understand I am writing to a *human being*, doing this open source project on his *free* time
- [x] I will be involved in my own issue, by answering questions and informing if the issue is resolved

# Configuration

- VS Code version: 1.53.0
- Angular version (`ng version`):
```
Angular CLI: 11.1.2
Node: 12.18.4
OS: win32 x64

Angular: 11.1.1
... animations, cdk, common, compiler, compiler-cli, core, forms
... language-service, material, platform-browser
... platform-browser-dynamic, router
Ivy Workspace: Yes

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1101.2
@angular-devkit/build-angular   0.1101.2
@angular-devkit/core            11.1.2
@angular-devkit/schematics      11.1.2
@angular/cli                    11.1.2
@angular/flex-layout            11.0.0-beta.33
@schematics/angular             11.1.2
@schematics/update              0.1101.2
rxjs                            6.6.3
typescript                      4.1.3
```

# Description of the Feature

It is a pretty common practice in Angular to use `index.ts` (barrel) files to re-export services/components/etc to make importing them easier in other areas of the application.

For example, I may have a `services` folder with a `user.service.ts` and a `permissions.service.ts`. Rather than import them separately in a component that needs both services, I generally export both services in the `services/index.ts` file like so:

```
export { UserService } from './user.service';
export { PermissionsService } from './permissions.service';
```

Some people may prefer to use the `*` method instead, so perhaps that could be an option as well for this feature:

```
export * from './user.service';
export * from './permissions.service';
```

Both methods of exporting allow me to easily import both services in a component (ex: `components/profile.component.ts`) like so:

```
import { UserService, PermissionsService } from '../services';
```

# Additions to the Extension

1. An option to `export { generated-item } from generated-item-file` in `index.ts` (if one exists in the context)
2. An option to import the `generated-item` in the module after step 1 (that way it is imported via barrel file)
3. An option to `export * from generated-item-file` if that line doesn't already exist in the `index.ts` (if one exists in the context)
4. (maybe?) An option to create the `index.ts` if none exists

Hopefully I was clear in explaining what I would like to have the extension automate for me. Happy to explain further if I wasn't clear enough.

Thanks!

**Comment by cyrilletuzi** [comment]
Hi again,

This is an idea I already have for quite some time, and there would be many other great ideas for enhancing generation like this one.

Note however that the extension is just a graphic interface for existing schematics, it does not generate anything by itself. While it would not be impossible to do generation directly inside the extension code, it would clearly be better in a separate tool, ie. schematics, that could complement the extension (or even be included directly in the extension, but still a separate tool).

The main issue is time, or more exactly money. This extension is already hundreds hours of unpaid work, all by myself (and I have another popular Angular tool to maintain). I've reached the limit to what I can do unpaid, and currently I only have one [sponsor](https://github.com/sponsors/cyrilletuzi/).

Your GitHub profile suggests that you work on your own, so I can't ask you if your company would sponsor this work, but clearly with the adequate support I could make a whole architecture tool, including this idea.

Moving this to the new GitHub "Discussions" tab (where you can continue the discussion), as it is not an issue.

**Comment by cyrilletuzi** [comment]
Hi @andy-bond, for information, this is available in the [Pro edition](https://cyrilletuzi.gumroad.com/l/schematicspro/).
